misc/python/materialize/mzbuild.py

```diff
@@ -160,21 +160,38 @@ def extra(self) -> str:
         return ""
 
 
-class Directory(PreImage):
-    """ A `PreImage` action where all files in a directory are defined as inputs."""
+class Copy(PreImage):
+    """A `PreImage` action which copies files from a directory.
+
+    The contents of the specified `source` directory are copied into the
+    `destination` directory. The `source` directory is relative to the
+    repository root. The `destination` directory is relative to the mzbuild
+    context. Files in the `source` directory are matched against the glob
+    specified by the `matching` argument.
+    """
 
     def __init__(self, rd: RepositoryDetails, path: Path, config: Dict[str, Any]):
         super().__init__(rd, path)
 
         self.source = config.pop("source", None)
-
         if self.source is None:
             raise ValueError("mzbuild config is missing 'source' argument")
-        else:
-            self.source = self.rd.root / self.source
+
+        self.destination = config.pop("destination", None)
+        if self.destination is None:
+            raise ValueError("mzbuild config is missing 'destination' argument")
+
+        self.matching = config.pop("matching", "*")
+
+    def run(self) -> None:
+        super().run()
+        for src in self.inputs():
+            dst = self.path / self.destination / Path(src).relative_to(self.source)
+            dst.parent.mkdir(parents=True, exist_ok=True)
```

Comment:
Given that `super().run()` is running `git clean`, should we bail the `destination` directory exists?

Reply:
It's not quite that simple. If you're copying files `src/foo/a` and `src/foo/b` then we'll call mkdir on `src/foo` twice. We could definitely try to be more clever, but not sure it's worth it.

Reply:
Right, because this is the parent directory. Makes sense to me, thanks!